app/src/main/java/ch/blinkenlights/android/vanilla/playback/PlaybackTimestampHandler.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) 2019 Felix Nüsse <felix.nuesse@t-online.de>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package ch.blinkenlights.android.vanilla.playback;
+
+import android.content.Context;
+import android.os.Handler;
+import android.util.Log;
+
+import ch.blinkenlights.android.medialibrary.MediaLibrary;
+import ch.blinkenlights.android.medialibrary.MediaLibraryBackend;
+import ch.blinkenlights.android.vanilla.Song;
+import ch.blinkenlights.android.vanilla.VanillaMediaPlayer;
+
+public class PlaybackTimestampHandler {
+
+	final Handler mHandler = new Handler();
+
+	private MediaLibraryBackend mlb;
+	private Context c;
+
+	private boolean alreadyUpdating=false;
+	private Song mSong;
+	private int mTimestamp;
+
+	private static String TAG = "PlaybackUpdateHandler";
+
+	public PlaybackTimestampHandler(Context c) {
+		this.c = c;
+	}
+
+	public void stopUpdates(){
+		Log.d(TAG, "Called on main thread stopUpdates");
+		mHandler.removeCallbacksAndMessages(null);
+		alreadyUpdating=false;
+	}
+
+	public void start(VanillaMediaPlayer mediaPlayer) {
+
+		if(!mediaPlayer.isPlaying()){
+			//Log.d(TAG, "Mediaplayer not playing!");
+			//return;
+		}
+
+		if(alreadyUpdating){
+			//Log.d(TAG, "Mediaplayer already updating!");
+			return;
+		}
+
+		final VanillaMediaPlayer vmp = mediaPlayer;
+		// Create the Handler object (on the main thread by default)
+
+		// Define the code block to be executed
+		Runnable runnableCode = new Runnable() {
+			@Override
+			public void run() {
+				try {
+					//Log.d(TAG, "Timestamp: "+vmp.getCurrentPosition()+" - "+mSong.title+" - "+mSong.id);
+					mTimestamp = vmp.getCurrentPosition();
+					storeTimestampdataForSong();
+				} finally {
+					//if mediaplayer is not playing, wait longer.
+					int delay=500;
+					if(!vmp.isPlaying()){
+						delay=5000;
+						//Log.d(TAG, "Long Delay: "+delay);
+					}
+					mHandler.postDelayed(this, delay);
```

Comment:
Wait: so this code performs an SQL upadte ever 500 / 5000ms?

That's not a cheap operation and will have significant effects on the battery consumption of the app.
This must be event driven: polling the play state is just completely wrong.

Reply:
How can this be event driven? As far as i know the mediaplayer does not provide a callback, so that polling is the only way (The documentation also suggests to use the getCurrentPosition() call to keep track of the playback progress)

Albeit, i have adjusted the code to stop updating when the playback is stopped, and increased the normal delay to 2 seconds, which at least reduces the load by 75%

What we can do additionally is to store the value not directly in the database, but in a shared preference every two seconds, and only update the database value every minute (and on application restart if it crashed for whatever reason, and always before we start a new track) but i am not sure if that is computationally more easy than just writing to the database

Reply:
> As far as i know the mediaplayer does not provide a callback

You want to save the position if the song was paused or changed: vanilla has hooks for that.